### PR TITLE
CI: Retrieve pyright version from requirements file

### PR DESCRIPTION
**Pull request by webmiche**
Okay, this is done now. The pyright workflows now first parse the requirements file and write the proper pyright version to a file. This file is then read in as an environment variable, which is later used to decide the version.

This means we can now just merge the bumps that dependabot suggests.

### CODE REVIEW

**Comment by webmiche** [comment]
Okay, this is done now. The pyright workflows now first parse the requirements file and write the proper pyright version to a file. This file is then read in as an environment variable, which is later used to decide the version.

This means we can now just merge the bumps that dependabot suggests.

**Review by tobiasgrosser — APPROVED** [review]
How nice is this! Thank you for working on this. This is a great solution to counteract my overeager merging.

File: requirements-optional.txt [hunk 1]
```diff
@@ -3,5 +3,6 @@ toml<0.11
 pytest-cov
 coverage<8.0.0
 ipykernel
-# Remember to update the CI pyright versions when updating this line
+# pyright has to be the last line and fixed with `==`. The CI parses this file
```
Comment (superlopuh) [inline requirements-optional.txt:6]: nit: would be good to mention the path of the file that parses it

File: parse_and_set_pyright_version.py
Comment (superlopuh) [inline parse_and_set_pyright_version.py:1]: nit: could move this to .github maybe?

**Review by superlopuh — APPROVED** [review]
Very nice

**Review by georgebisbas — COMMENTED** [review]
Looks like adding a lot of code, to only;y avoid checking more major versions ;-)

**Comment by webmiche** [comment]
> Looks like adding a lot of code, to only;y avoid checking more major versions ;-)

I agree, this is pretty much a bandaid on top of a bandaid. I am not even sure we want this kind of thing in the repo. Still, having a breaking CI on main every now and then is even more annoying, tbh.

IMO, we could also just go with fixed pyright versions and not upgrade it, unless we really want something that was recently added, in which case we update by hand.

This would mean we pretty much handle it the same as the MLIR version currently, which is fine for me.

**Comment by superlopuh** [comment]
I don't remember this for sure but I think Pylance, which is updated in lockstep with Pyright, shows the errors for the newer linter, so it's good for us to update it automatically too.

File: parse_and_set_pyright_version.py
Comment (math-fehr) [reply parse_and_set_pyright_version.py:1]: Agreed, so it won't pollute the root!

**Review by math-fehr — APPROVED** [review]
Nice!
I think it's good if we can stay up to date with pyright, but I'm also fine in updating it less, if people feel strongly against it.
In any case, we can just refuse dependabot patch if we don't like it, right?

**Pull request merged**